pkg/slices/slices.go

```diff
@@ -18,20 +18,84 @@ limitations under the License.
 
 package slices
 
+import "reflect"
+
+// Intersect returns the intersection of two slices using the given key function.
+// The returned intersection preserves the order of elements in s2.
 func Intersect[E any, K comparable](s1, s2 []E, key func(E) K) []E {
-	m := make(map[K]bool)
+	var r []E
 
-	for _, v := range s1 {
-		m[key(v)] = true
+	m := mapFrom(s1, key)
+	for i := range s2 {
+		if _, ok := m[key(s2[i])]; ok {
+			r = append(r, s2[i])
+		}
 	}
 
-	var r []E
+	return r
+}
 
-	for _, v := range s2 {
-		if _, ok := m[key(v)]; ok {
-			r = append(r, v)
+// IndexOf returns the index of the first occurrence of element e in s using the given key function, or -1 if not present.
+func IndexOf[E any, K comparable](s []E, e K, key func(E) K) int {
+	for i := range s {
+		if key(s[i]) == e {
+			return i
 		}
 	}
 
-	return r
+	return -1
+}
+
+// AppendIfNotPresent appends element e to s if not already present using the given key function. Returns the resultant
+// slice and a bool indicating if the element was added.
+func AppendIfNotPresent[E any, K comparable](s []E, e E, key func(E) K) ([]E, bool) {
+	i := IndexOf(s, key(e), key)
+	if i >= 0 {
+		return s, false
+	}
+
+	return append(s, e), true
+}
+
+// Remove the first instance of element e from s using the given key function. Returns the resultant slice and a bool
+// indicating if the element was removed.
+func Remove[E any, K comparable](s []E, e E, key func(E) K) ([]E, bool) {
+	i := IndexOf(s, key(e), key)
+	if i < 0 {
+		return s, false
+	}
+
+	copy(s[i:], s[i+1:])
+
+	return s[:len(s)-1], true
+}
+
+// Union returns the union of two slices using the given key function.
+func Union[E any, K comparable](s1, s2 []E, key func(E) K) []E {
+	m := mapFrom(s1, key)
+	for i := range s2 {
+		if _, ok := m[key(s2[i])]; !ok {
+			s1 = append(s1, s2[i])
+		}
+	}
+
+	return s1
+}
+
+// Equivalent reports whether two slices have the same length and contain the same elements regardless of order and
+// uniqueness.
```

Comment:
Length no longer matters (if `s1` contains two elements resulting in the same key, and `s2` contains it once, they will have different lengths but be equivalent):

```suggestion
// Equivalent reports whether two slices contain the same elements regardless of order and
// uniqueness.
```